### ISSUE TITLE
Exception  thrown during recovery

### ISSUE THREAD

**Issue by mstaros** [microsoft/FASTER#393]
What does  this error mean?

 System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at FASTER.core.VariableLengthBlittableAllocator`2.ClearPage(Int64 page, Int32 offset)
   at FASTER.core.AllocatorBase`2.RecoveryReset(Int64 tailAddress, Int64 headAddress, Int64 beginAddress, Int64 readonlyAddress)
   at FASTER.core.FasterKV`2.InternalRecover(IndexCheckpointInfo recoveredICInfo, HybridLogCheckpointInfo recoveredHLCInfo, Int32 numPagesToPreload, Boolean undoFutureVersions)
   at FASTER.core.FasterKV`2.InternalRecover(Guid indexToken, Guid hybridLogToken, Int32 numPagesToPreload, Boolean undoFutureVersions)
   at FASTER.core.FasterKV`2.Recover(Guid indexCheckpointToken, Guid hybridLogCheckpointToken, Int32 numPagesToPreload, Boolean undoFutureVersions)

**Comment by badrishc** [comment]
Can you describe the sequence of events that caused this error.  Also, what kind of checkpoints were you taking. And did it involve index resizing? Any detailed info or repro would be useful.

**Comment by mstaros** [comment]
> Can you describe the sequence of events that caused this error. Also, what kind of checkpoints were you taking. And did it involve index resizing? Any detailed info or repro would be useful.

There was no growIndex command. Type of checkpoint is foldover. It has happened again on newly created KV after several dozens of "saves" and "recoveries".

**Comment by badrishc** [comment]
@mstaros  - you cannot just increase the size of index in constructor without growing the index, because index resizing involves recomputing the hash codes.

Good news is that we have now fixed `GrowIndex` to work correctly. It will double the index size and return only after the index doubling is complete. It should work with checkpoints as well. See example code at https://github.com/microsoft/FASTER/blob/c7da0eba21fcc923d636a5644b8509f32af0af24/cs/test/RecoveryChecks.cs#L341

**Comment by mstaros** [comment]
> @mstaros - you cannot just increase the size of index in constructor without growing the index, because index resizing involves recomputing the hash codes.
> 
> Good news is that we have now fixed `GrowIndex` to work correctly. It will double the index size and return only after the index doubling is complete. It should work with checkpoints as well. See example code at
> 
> https://github.com/microsoft/FASTER/blob/c7da0eba21fcc923d636a5644b8509f32af0af24/cs/test/RecoveryChecks.cs#L341


It looks much more resizable now)  But then it would be nice to have a new  index size after GrowIndex( as out parameter may be).
Also what will happen if index size in ctor is increased?   and how this value will correlate with new index size after GrowIndex?  and when the exception will be thrown?

**Comment by badrishc** [comment]
You can query `store.IndexSize` to get the size of the index after the `GrowIndex` command. It will return the size of index in #cache lines (64 bytes are used per cache line). In the linked PR, we have added an API to also get the space used by overflow buckets (again, 64 bytes each): `store.OverflowBucketCount`. Is this sufficient?

Index size specified in constructor is the initial size before recovery. If you then recover from an existing checkpoint with a hash table of a different size, we will replace the current index with the appropriately sized one during recovery. So, no exception will be thrown. Hope that helps.

**Comment by mstaros** [comment]
> You can query `store.IndexSize` to get the size of the index after the `GrowIndex` command. It will return the size of index in #cache lines (64 bytes are used per cache line). In the linked PR, we have added an API to also get the space used by overflow buckets (again, 64 bytes each): `store.OverflowBucketCount`. Is this sufficient?
> 
Yes, this is sufficient of course.
Should GrowIndex be called only when all sessions are suspended?  If it is the case - if there is any such API or one has to manage sessions' suspensions manually?

**Comment by badrishc** [comment]
`GrowIndex` has been designed to work concurrently with user session operations. However,  we do not yet exercise this concurrency in testcases. Let us know if you run into any issues with it.
